finality-aleph/src/testing/network.rs

```diff
@@ -321,3 +489,227 @@ async fn test_connects_to_others() {
     test_data.cleanup().await;
     assert_eq!(data_network.next().await, None);
 }
+
+#[tokio::test]
+async fn test_connects_to_others_early_validator() {
+    let session_id = 43;
+    let mut test_data = prepare_one_session_test_data().await;
+    test_data.early_start_validator_session(0, session_id);
+    test_data.connect_session_authorities(session_id).await;
+    test_data.check_sends_add_reserved_node().await;
+    test_data
+        .check_sends_authentication(
+            test_data
+                .get_session_handler(0, session_id)
+                .await
+                .authentication()
+                .unwrap(),
+        )
+        .await;
+    let mut data_network = test_data.start_validator_session(0, session_id).await;
+
+    let data = vec![1, 2, 3];
+    test_data.emit_notifications_received(
+        1,
+        vec![NetworkData::Data(data.clone(), SessionId(session_id))],
+    );
+    assert_eq!(
+        timeout(DEFAULT_TIMEOUT, data_network.next()).await,
+        Ok(Some(data.clone()))
+    );
+
+    test_data.cleanup().await;
+    assert_eq!(data_network.next().await, None);
+}
+
+#[tokio::test]
+async fn test_stops_session() {
+    let session_id = 43;
+    let mut test_data = prepare_one_session_test_data().await;
+    let mut data_network = test_data.start_session(session_id).await;
+
+    test_data
+        .session_manager
+        .stop_session(SessionId(session_id))
+        .unwrap();
+    assert_eq!(
+        timeout(DEFAULT_TIMEOUT, test_data.network.remove_reserved.next())
+            .await
+            .ok()
+            .flatten(),
+        Some((
+            HashSet::from_iter(test_data.authorities.iter().skip(1).map(|a| a.peer_id())),
+            Protocol::Validator.name()
+        ))
+    );
+
+    // This assert should be before cleanup
```

Comment:
Why?

Reply:
We want to check whether  `session_manager.stop_session(...)` drops the sender. After cleanup all network tasks end and senders will be dropped, so if assert was after cleanup we wouldn't know whether `data_network` receiver is droopped because of `.stop_session(...)` or because of cleanup.

Reply:
Consider extending that comment with this explanation.